common/config/azure-pipelines/templates/publish.yaml

```diff
@@ -18,37 +18,44 @@ steps:
      import sys, os, glob, re, subprocess
      import shutil
 
-     def determineDistTag(currVersion):
+     def determineDistTag(currentVer, latestVer):
        ## The master branch is the only one that should get the 'nightly' release tag
        mainBranch = "master"
 
-       # Branch that is the current set of releases, will be given the 'latest' dist tag
-       currentReleaseBranch = "release/2."
-
-       # Branches that used to hold the current releases but will no longer now not be given the 'latest' dist tag
-       # Note: This should probably be an array in the future.
-       oldReleaseBranch = "release/1."
-
        distTag = None
        if len(sys.argv) == 4:
          branchName = sys.argv[3]
 
-         print ("Branch name: " + branchName + " and current version " + currVersion)
+         print ("Branch name: " + branchName + "\nCurrent version: " + currentVer + "\nLatest version: " + latestVer)
 
          # The most common case is the tag will be a nightly tag
          if mainBranch in branchName:
            distTag = "nightly"
          elif "release/" in branchName:
            print ("On a release branch")
-           if currentReleaseBranch in branchName:
-             distTag = "latest"
-           elif oldReleaseBranch in branchName:
-             distTag = "previous"
-
-           # Override the 2 previous checks for the case where we're releasing a -dev version off a release branch.
-           # These will always be considered rc.
-           if "-" in currVersion:
+           currentDevVer = -1
+           if ("-" in currentVer):
+             currentVer, currentDevVer = currentVer.split("-")
+             currentDevVer = int(currentDevVer.split(".")[1])
+
+           currentMajorVer, currentMinorVer, currentPatchVer = currentVer.split(".")
+           latestMajorVer, latestMinorVer, latestPatchVer = latestVer.split(".")
+           currentMajorVer, currentMinorVer, currentPatchVer = int(currentMajorVer), int(currentMinorVer), int(currentPatchVer)
+           latestMajorVer, latestMinorVer, latestPatchVer = int(latestMajorVer), int(latestMinorVer), int(latestPatchVer)
+
+           if (currentDevVer != -1):
              distTag = "rc"
+           else:
+             if currentMajorVer > latestMajorVer:
+               distTag = "latest"
+             elif currentMajorVer < latestMajorVer:
+               distTag = "previous"
+             else:
+               if currentMinorVer < latestMinorVer:
+                 distTag = "previous"
```

Comment:
I'm debating how we want to handle this case, the major version workflow is correct since we want to apply the `previous` tag to all versions released on any old major version. However, in the case for minor versions we still want the latest major version to always have the `latest` tag and the previous major version to have the `previous`.

For example, if we have `2.19.20` = `previous`, `3.1.0` = `latest` and we are currently trying to release a `3.0.1`. This workflow would make `3.0.1` = `previous` and `2.19.20` would lose all tags. I think we'd rather have `2.19.20` keep its tag as `previous` and need `3.1.0` to still be latest. Then in an ideal world the `3.0.1` release would be without a tag at all.

Reply:
I've changed it so now any new release on an older major version will get the 'previous' dist-tag, and a new patch version on an older release branch of the current major version (e.g. 3.0.1 is released when the latest release is 3.1.0) will receive no dist-tag. 

I was also thinking that it might make sense to only apply the 'previous' tag to the newest release on the previous major branch (if the current 'previous' release is 2.19.21 for example, 2.18.5 or 1.x would receive no tag while 2.19.22 or 2.20 would receive the 'previous' tag).

Reply:
I agree with that statement, the ideal scenario is that we'd never even be doing a patch for any old major versions with not the latest minor version of that major... However, it has happened before so we should probably account for it.

Reply:
I've just pushed some new changes so now we should be following that behavior ('previous' tag is only added if it's the latest release of the previous major version).